sql: TestLogic/fakedist-disk/aggregate/string_agg is flaky

Running `make stress TESTS='TestLogic' PKG=./pkg/sql/logictest TESTFLAGS='-config fakedist-disk -d testdata/logic_test/aggregate'` on c783185e39c3046e114e54ebe7df56da2291edfb produces:
```
            --- FAIL: TestLogic/fakedist-disk/aggregate/string_agg (0.61s)
                logic.go:1706:
                    testdata/logic_test/aggregate:2080: SELECT a, b, count(*) FROM ab RIGHT JOIN tab ON b=col2 GROUP BY a
                    expected:
                        NULL  NULL  4
                        4     7     2
                    but found (query options: "rowsort" -> ignore the following ordering of rows) :
                        NULL  NULL  2
                        NULL  NULL  2
                        4     7     2
```

--- a/pkg/sql/colexec/aggregator_test.go
+++ b/pkg/sql/colexec/aggregator_test.go
@@ -34,12 +34,13 @@ var (
 type aggregatorTestCase struct {
 	// colTypes, aggFns, groupCols, and aggCols will be set to their default
 	// values before running a test if nil.
-	colTypes  []coltypes.T
-	aggFns    []execinfrapb.AggregatorSpec_Func
-	groupCols []uint32
-	aggCols   [][]uint32
-	input     tuples
-	expected  tuples
+	colTypes       []coltypes.T
+	aggFns         []execinfrapb.AggregatorSpec_Func
+	groupCols      []uint32
+	aggCols        [][]uint32
+	input          tuples
+	unorderedInput bool
+	expected       tuples
 	// {output}BatchSize() if not 0 are passed in to NewOrderedAggregator to
 	// divide input/output batches.
 	batchSize       int
@@ -271,6 +272,19 @@ func TestAggregatorOneFunc(t *testing.T) {
 			groupCols:       []uint32{1, 2},
 			aggCols:         [][]uint32{{0}},
 		},
+		{
+			input: tuples{
+				{nil, 1},
+				{4, 42},
+				{nil, 2},
+			},
+			expected: tuples{
+				{3},
+				{42},
+			},
+			name:           "UnorderedWithNullsInGroupingCol",
+			unorderedInput: true,
+		},
 	}
 
 	// Run tests with deliberate batch sizes and no selection vectors.
@@ -280,31 +294,38 @@ func TestAggregatorOneFunc(t *testing.T) {
 				t.Fatal(err)
 			}
 
-			tupleSource := newOpTestInput(tc.batchSize, tc.input, nil /* typs */)
-			a, err := NewOrderedAggregator(
-				testAllocator,
-				tupleSource,
-				tc.colTypes,
-				tc.aggFns,
-				tc.groupCols,
-				tc.aggCols,
-				false, /* isScalar */
-			)
-			if err != nil {
-				t.Fatal(err)
-			}
+			if !tc.unorderedInput {
+				tupleSource := newOpTestInput(tc.batchSize, tc.input, nil /* typs */)
+				a, err := NewOrderedAggregator(
+					testAllocator,
+					tupleSource,
+					tc.colTypes,
+					tc.aggFns,
+					tc.groupCols,
+					tc.aggCols,
+					false, /* isScalar */
+				)
+				if err != nil {
+					t.Fatal(err)
+				}
 
-			out := newOpTestOutput(a, tc.expected)
-			// Explicitly reinitialize the aggregator with the given output batch
-			// size.
-			a.(*orderedAggregator).initWithInputAndOutputBatchSize(tc.batchSize, tc.outputBatchSize)
-			if err := out.VerifyAnyOrder(); err != nil {
-				t.Fatal(err)
+				out := newOpTestOutput(a, tc.expected)
+				// Explicitly reinitialize the aggregator with the given output batch
+				// size.
+				a.(*orderedAggregator).initWithInputAndOutputBatchSize(tc.batchSize, tc.outputBatchSize)
+				if err := out.VerifyAnyOrder(); err != nil {
+					t.Fatal(err)
+				}
 			}
 
 			// Run randomized tests on this test case.
 			t.Run(fmt.Sprintf("Randomized"), func(t *testing.T) {
 				for _, agg := range aggTypes {
+					if tc.unorderedInput && agg.name == "ordered" {
+						// This test case has unordered input, so we skip ordered
+						// aggregator.
+						continue
+					}
 					t.Run(agg.name, func(t *testing.T) {
 						runTests(t, []tuples{tc.input}, tc.expected, unorderedVerifier,
 							func(input []Operator) (Operator, error) {

--- a/pkg/sql/colexec/hash_aggregator_tmpl.go
+++ b/pkg/sql/colexec/hash_aggregator_tmpl.go
@@ -72,8 +72,11 @@ func _MATCH_LOOP(
 	for selIdx, rowIdx := range sel {
 		// {{if .LhsMaybeHasNulls}}
 		// {{if .RhsMaybeHasNulls}}
-		diffNull := (lhsNull != rhs.Nulls().NullAt(rowIdx))
-		if diffNull {
+		rhsNull := rhs.Nulls().NullAt(rowIdx)
+		if lhsNull && rhsNull {
+			// Both values are NULLs, and we do not consider them different.
+			continue
+		} else if lhsNull || rhsNull {
 			diff[selIdx] = true
 			continue
 		}
@@ -126,7 +129,6 @@ func (v hashAggFuncs) match(
 	for keyIdx, colIdx := range keyCols {
 		lhs := keyMapping.ColVec(keyIdx)
 		lhsHasNull := lhs.MaybeHasNulls()
-		lhsNull := lhs.Nulls().NullAt(v.keyIdx)
 
 		rhs := b.ColVec(int(colIdx))
 		rhsHasNull := rhs.MaybeHasNulls()
@@ -138,14 +140,19 @@ func (v hashAggFuncs) match(
 		case _TYPES_T:
 			lhsCol := lhs._TemplateType()
 			rhsCol := rhs._TemplateType()
-			if lhsHasNull && rhsHasNull {
-				_MATCH_LOOP(sel, lhs, rhs, aggKeyIdx, lhsNull, diff, true, true)
-			} else if lhsHasNull && !rhsHasNull {
-				_MATCH_LOOP(sel, lhs, rhs, aggKeyIdx, lhsNull, diff, true, false)
-			} else if !lhsHasNull && rhsHasNull {
-				_MATCH_LOOP(sel, lhs, rhs, aggKeyIdx, lhsNull, diff, false, true)
+			if lhsHasNull {
+				lhsNull := lhs.Nulls().NullAt(v.keyIdx)
+				if rhsHasNull {
+					_MATCH_LOOP(sel, lhs, rhs, aggKeyIdx, lhsNull, diff, true, true)
+				} else {
+					_MATCH_LOOP(sel, lhs, rhs, aggKeyIdx, lhsNull, diff, true, false)
+				}
 			} else {
-				_MATCH_LOOP(sel, lhs, rhs, aggKeyIdx, lhsNull, diff, false, false)
+				if rhsHasNull {
+					_MATCH_LOOP(sel, lhs, rhs, aggKeyIdx, lhsNull, diff, false, true)
+				} else {
+					_MATCH_LOOP(sel, lhs, rhs, aggKeyIdx, lhsNull, diff, false, false)
+				}
 			}
 		// {{end}}
 		default:

